clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/client/OpenstackClientProvider.groovy

```diff
@@ -52,18 +58,65 @@ abstract class OpenstackClientProvider {
   }
 
   /**
-   * Handler for an openstack4j request.
-   * @param closure
-   * @return
+   * Create or update a security group, applying a list of rules. If the securityGroupId is provided, updates an existing
+   * security group, else creates a new security group.
+   *
+   * Note: 2 default egress rules are created when creating a new security group
+   * automatically with remote IP prefixes 0.0.0.0/0 and ::/0.
+   *
+   * @param securityGroupId id of an existing security group to update
+   * @param securityGroupName name security group
+   * @param description description of the security group
+   * @param rules list of rules for the security group
+   */
+  void upsertSecurityGroup(String securityGroupId, String securityGroupName, String description, List<OpenstackSecurityGroupDescription.Rule> rules) {
+
+     handleRequest(AtomicOperations.UPSERT_SECURITY_GROUP) {
+
+       // The call to getClient reauthentictes via a token, so grab once for this method
```

Comment:
I was pointing out that `getClient` is not a simple getter, so I only wanted to call it once for this thread.